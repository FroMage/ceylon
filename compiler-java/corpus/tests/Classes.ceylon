class Classes() {

	class Simple() {}
	
	class WithMethods() {
		void doNothing() {}
		Natural returnZero() { return 0; }
		String returnArgument(String arg) { return arg; }
	}
	
	class WithAttributes() {
		Natural count = 0;
		mutable String description := "";
		String countAsString { return $count; }
		assign countAsString { count := countAsString.parseNatural; }
	}
	
	abstract class WithAbstractMethods() {
		void doNothing();
		Natural returnZero();
		String returnArgument(String arg);
	}
	
	abstract class WithAbstractAttributes() {
		Natural count;
		mutable String description;
	}
	
	class WithInitParameters(Natural count, String description) {
		Natural count = count;
		mutable String description := description;
	}
	
	doc "This class has annotations. Remember that literal
	     strings can be spread across multiple lines."
	by "Gavin King"
	   "Andrew Haley"
	scope(session)
	see #Simple
	throws #DatabaseException
	       "if database access fails"
	entity table { name="someTable"; schema="someSchema"; }
	class WithAnnotations() {}
	
	class Extends() extends Simple() {}
	
	class ExtendsWithInitParameters(Natural count, String description, Float tolerance) 
			extends WithInitParameters(count, description) {
		Float tol = tolerance;
	}
	
	class WithClosedInstanceList() { case foo, case bar, case baz; }
	
	class WithOpenInstanceList() { case foo, case bar, case baz... }
	
	class WithInstanceListAndInitParameter(Integer i) {
		case one(1), case two(2), case three(3); 
		Integer value = i;
	}
	
	class WithInstanceListAndFunctionalInitParameter(String name()) {
		
		case foo { String name() { return "Foo" } }, 
		case bar { String name() { return "Bar" } },
		case baz { String name() { return "Baz" } };
		
		String name() = name;
		
	}
	
	class WithAnnotatedInstanceList(Integer i) {
		doc "1" numericValue(1) case one(1), 
		doc "2" numericValue(2) case two(2), 
		doc "3" numericValue(3) case three(3); 
		Integer value = i;
	}
	
	class WithTypeParameters<X, Y>() {}
	
	class WithTypeConstraints<X, Y>()
		where X satisfies String 
		where Y(Natural count) {}
		
	class WithNestedClass() {
		class NestedClass() {}
	}
	
	class WithNestedInterface() {
		interface NestedInterface {}
	}
	
	public interface Counter {
		public Natural count;
		public void inc();
	}
	
	public interface Resettable {
		void reset();
	}

	class Satisfies() satisfies Counter {
		mutable Natural count := 0;
		void inc() { count++; }
	}
	
	class SatisfiesMultiple() satisfies Counter, Resettable {
		mutable Natural count := 0;
		void inc() { count++; }
		void reset() { count := 0; }
	}
	
}