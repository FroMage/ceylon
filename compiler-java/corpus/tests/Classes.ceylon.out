# Classes.ceylon

(TYPE_DECL
  (TYPE_DECL
    (CLASS_DECL
      (TYPE_NAME Classes)
      (STMT_LIST
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME Simple) STMT_LIST))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithMethods)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME doNothing) STMT_LIST)
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME returnZero)
                (STMT_LIST
                  (RET_STMT
                    (EXPR
                      (INT_CST 0)))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME returnArgument)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME arg)))
                (STMT_LIST
                  (RET_STMT
                    (EXPR
                      (MEMBER_NAME arg))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithAttributes)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME count)
                (INIT_EXPR
                  (EXPR
                    (INT_CST 0))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME description)
                (INIT_EXPR
                  (EXPR
                    (STRING_CST "")))
                (ANNOTATION_LIST
                  (LANG_ANNOTATION mutable)))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME countAsString)
                (STMT_LIST
                  (RET_STMT
                    (EXPR
                      (PREFIX_EXPR $
                        (MEMBER_NAME count))))))
              (MEMBER_DECL
                (ATTRIBUTE_SETTER
                  (MEMBER_NAME countAsString))
                (STMT_LIST
                  (EXPR
                    (:=
                      (MEMBER_NAME count)
                      (EXPR
                        (EXPR
                          (MEMBER_NAME countAsString) DOT
                          (MEMBER_NAME parseNatural))))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithAbstractMethods)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME doNothing))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME returnZero))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME returnArgument)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME arg))))))
          (ANNOTATION_LIST
            (LANG_ANNOTATION abstract)))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithAbstractAttributes)
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME count))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME description)
                (ANNOTATION_LIST
                  (LANG_ANNOTATION mutable)))))
          (ANNOTATION_LIST
            (LANG_ANNOTATION abstract)))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithInitParameters)
            (FORMAL_PARAMETER_LIST
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME Natural))
                (ARG_NAME count))
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME String))
                (ARG_NAME description)))
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME count)
                (INIT_EXPR
                  (EXPR
                    (MEMBER_NAME count))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME description)
                (INIT_EXPR
                  (EXPR
                    (MEMBER_NAME description)))
                (ANNOTATION_LIST
                  (LANG_ANNOTATION mutable))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithAnnotations) STMT_LIST)
          (ANNOTATION_LIST
            (USER_ANNOTATION
              (ANNOTATION_NAME doc)
              (STRING_CST "This class has annotations. Remember that literal
	     strings can be spread across multiple lines."))
            (USER_ANNOTATION
              (ANNOTATION_NAME by)
              (STRING_CST "Andrew Haley"))
            (USER_ANNOTATION
              (ANNOTATION_NAME scope)
              (ARG_LIST
                (EXPR
                  (MEMBER_NAME session))))
            (USER_ANNOTATION
              (ANNOTATION_NAME see))
            (USER_ANNOTATION
              (ANNOTATION_NAME throws)
              (STRING_CST "if database access fails"))
            (USER_ANNOTATION
              (ANNOTATION_NAME entity))
            (USER_ANNOTATION
              (ANNOTATION_NAME table)
              (ARG_LIST
                (NAMED_ARG
                  (ARG_NAME name)
                  (INIT_EXPR
                    (EXPR
                      (STRING_CST "someTable"))))
                (NAMED_ARG
                  (ARG_NAME schema)
                  (INIT_EXPR
                    (EXPR
                      (STRING_CST "someSchema"))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME Extends)
            (SUPERCLASS
              (TYPE
                (TYPE_NAME Simple)) ARG_LIST) STMT_LIST))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME ExtendsWithInitParameters)
            (FORMAL_PARAMETER_LIST
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME Natural))
                (ARG_NAME count))
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME String))
                (ARG_NAME description))
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME Float))
                (ARG_NAME tolerance)))
            (SUPERCLASS
              (TYPE
                (TYPE_NAME WithInitParameters))
              (ARG_LIST
                (EXPR
                  (MEMBER_NAME count))
                (EXPR
                  (MEMBER_NAME description))))
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Float)))
                (MEMBER_NAME tol)
                (INIT_EXPR
                  (EXPR
                    (MEMBER_NAME tolerance)))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithClosedInstanceList)
            (STMT_LIST
              (INST_DECL
                (MEMBER_NAME foo) ,)
              (INST_DECL
                (MEMBER_NAME bar) ,)
              (INST_DECL
                (MEMBER_NAME baz) ;))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithOpenInstanceList)
            (STMT_LIST
              (INST_DECL
                (MEMBER_NAME foo) ,)
              (INST_DECL
                (MEMBER_NAME bar) ,)
              (INST_DECL
                (MEMBER_NAME baz) ...))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithInstanceListAndInitParameter)
            (FORMAL_PARAMETER_LIST
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME Integer))
                (ARG_NAME i)))
            (STMT_LIST
              (INST_DECL
                (MEMBER_NAME one)
                (ARG_LIST
                  (EXPR
                    (INT_CST 1))) ,)
              (INST_DECL
                (MEMBER_NAME two)
                (ARG_LIST
                  (EXPR
                    (INT_CST 2))) ,)
              (INST_DECL
                (MEMBER_NAME three)
                (ARG_LIST
                  (EXPR
                    (INT_CST 3))) ;)
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Integer)))
                (MEMBER_NAME value)
                (INIT_EXPR
                  (EXPR
                    (MEMBER_NAME i)))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithInstanceListAndFunctionalInitParameter)
            (FORMAL_PARAMETER_LIST
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME String))
                (ARG_NAME name)))
            (STMT_LIST
              (INST_DECL
                (MEMBER_NAME foo)
                (ARG_LIST
                  (NAMED_ARG
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME name)
                    (STMT_LIST
                      (RET_STMT
                        (EXPR
                          (STRING_CST "Foo")))))) ,)
              (INST_DECL
                (MEMBER_NAME bar)
                (ARG_LIST
                  (NAMED_ARG
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME name)
                    (STMT_LIST
                      (RET_STMT
                        (EXPR
                          (STRING_CST "Bar")))))) ,)
              (INST_DECL
                (MEMBER_NAME baz)
                (ARG_LIST
                  (NAMED_ARG
                    (TYPE
                      (TYPE_NAME String))
                    (ARG_NAME name)
                    (STMT_LIST
                      (RET_STMT
                        (EXPR
                          (STRING_CST "Baz")))))) ;)
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME String)))
                (MEMBER_NAME name)
                (INIT_EXPR
                  (EXPR
                    (MEMBER_NAME name)))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithAnnotatedInstanceList)
            (FORMAL_PARAMETER_LIST
              (FORMAL_PARAMETER
                (TYPE
                  (TYPE_NAME Integer))
                (ARG_NAME i)))
            (STMT_LIST
              (INST_DECL
                (MEMBER_NAME one)
                (ARG_LIST
                  (EXPR
                    (INT_CST 1))) ,
                (ANNOTATION_LIST
                  (USER_ANNOTATION
                    (ANNOTATION_NAME doc)
                    (STRING_CST "1"))
                  (USER_ANNOTATION
                    (ANNOTATION_NAME numericValue)
                    (ARG_LIST
                      (EXPR
                        (INT_CST 1))))))
              (INST_DECL
                (MEMBER_NAME two)
                (ARG_LIST
                  (EXPR
                    (INT_CST 2))) ,
                (ANNOTATION_LIST
                  (USER_ANNOTATION
                    (ANNOTATION_NAME doc)
                    (STRING_CST "2"))
                  (USER_ANNOTATION
                    (ANNOTATION_NAME numericValue)
                    (ARG_LIST
                      (EXPR
                        (INT_CST 2))))))
              (INST_DECL
                (MEMBER_NAME three)
                (ARG_LIST
                  (EXPR
                    (INT_CST 3))) ;
                (ANNOTATION_LIST
                  (USER_ANNOTATION
                    (ANNOTATION_NAME doc)
                    (STRING_CST "3"))
                  (USER_ANNOTATION
                    (ANNOTATION_NAME numericValue)
                    (ARG_LIST
                      (EXPR
                        (INT_CST 3))))))
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Integer)))
                (MEMBER_NAME value)
                (INIT_EXPR
                  (EXPR
                    (MEMBER_NAME i)))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithTypeParameters)
            (TYPE_PARAMETER_LIST
              (TYPE_PARAMETER
                (TYPE_NAME X))
              (TYPE_PARAMETER
                (TYPE_NAME Y))) STMT_LIST))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithTypeConstraints)
            (TYPE_PARAMETER_LIST
              (TYPE_PARAMETER
                (TYPE_NAME X))
              (TYPE_PARAMETER
                (TYPE_NAME Y)))
            (TYPE_CONSTRAINT_LIST
              (TYPE_CONSTRAINT
                (TYPE_NAME X)
                (SATISFIES_LIST
                  (TYPE
                    (TYPE_NAME String))))
              (TYPE_CONSTRAINT
                (TYPE_NAME Y)
                (FORMAL_PARAMETER_LIST
                  (FORMAL_PARAMETER
                    (TYPE
                      (TYPE_NAME Natural))
                    (ARG_NAME count))))) STMT_LIST))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithNestedClass)
            (STMT_LIST
              (TYPE_DECL
                (CLASS_DECL
                  (TYPE_NAME NestedClass) STMT_LIST)))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME WithNestedInterface)
            (STMT_LIST
              (TYPE_DECL
                (INTERFACE_DECL
                  (TYPE_NAME NestedInterface))))))
        (TYPE_DECL
          (INTERFACE_DECL
            (TYPE_NAME Counter)
            (ABSTRACT_MEMBER_DECL
              (MEMBER_TYPE
                (TYPE
                  (TYPE_NAME Natural)))
              (MEMBER_NAME count)
              (ANNOTATION_LIST
                (LANG_ANNOTATION public)))
            (ABSTRACT_MEMBER_DECL
              (MEMBER_TYPE void)
              (MEMBER_NAME inc)
              (ANNOTATION_LIST
                (LANG_ANNOTATION public))))
          (ANNOTATION_LIST
            (LANG_ANNOTATION public)))
        (TYPE_DECL
          (INTERFACE_DECL
            (TYPE_NAME Resettable)
            (ABSTRACT_MEMBER_DECL
              (MEMBER_TYPE void)
              (MEMBER_NAME reset)))
          (ANNOTATION_LIST
            (LANG_ANNOTATION public)))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME Satisfies)
            (SATISFIES_LIST
              (TYPE
                (TYPE_NAME Counter)))
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME count)
                (INIT_EXPR
                  (EXPR
                    (INT_CST 0)))
                (ANNOTATION_LIST
                  (LANG_ANNOTATION mutable)))
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME inc)
                (STMT_LIST
                  (EXPR
                    (EXPR
                      (MEMBER_NAME count)
                      (POSTFIX_EXPR ++))))))))
        (TYPE_DECL
          (CLASS_DECL
            (TYPE_NAME SatisfiesMultiple)
            (SATISFIES_LIST
              (TYPE
                (TYPE_NAME Counter))
              (TYPE
                (TYPE_NAME Resettable)))
            (STMT_LIST
              (MEMBER_DECL
                (MEMBER_TYPE
                  (TYPE
                    (TYPE_NAME Natural)))
                (MEMBER_NAME count)
                (INIT_EXPR
                  (EXPR
                    (INT_CST 0)))
                (ANNOTATION_LIST
                  (LANG_ANNOTATION mutable)))
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME inc)
                (STMT_LIST
                  (EXPR
                    (EXPR
                      (MEMBER_NAME count)
                      (POSTFIX_EXPR ++)))))
              (MEMBER_DECL
                (MEMBER_TYPE void)
                (MEMBER_NAME reset)
                (STMT_LIST
                  (EXPR
                    (:=
                      (MEMBER_NAME count)
                      (EXPR
                        (INT_CST 0)))))))))))))
